Hdf5/feature restart process

jcotela opened the pull request:
A tool to initialize a simulation from an HDF5 output file. Cherry-picked from test-FIC-turbulent.

jcotela committed:
Making hdf5 input process open files in read_only mode by default.

jcotela committed:
Adding process to initialize simulation from HDF5 file.

sunethwarna commented on applications/HDF5Application/python_scripts/single_mesh_initialization_process.py:
```diff
@@ -0,0 +1,35 @@
+        "element_results_settings" : {},
+        "process_info_results_settings" : {},
+        "file_name": ""
+    }''')
```
old format of HDF5 settings

sunethwarna commented on applications/HDF5Application/python_scripts/single_mesh_initialization_process.py:
```diff
@@ -0,0 +1,35 @@
+    settings.ValidateAndAssignDefaults(default_settings)
+    model_part = Model[settings["model_part_name"].GetString()]
+    hdf5_file_factory = hdf5_io.HDF5SerialFileFactory(settings["file_settings"])
+    nodal_results_input = hdf5_io.NodalResultsInput(settings["nodal_results_settings"])
```
need to be modified according to new class "NodalSolutionStepDataInput".

msandre replied:
and NodalDataValueInput

sunethwarna commented on applications/HDF5Application/python_scripts/single_mesh_initialization_process.py:
```diff
@@ -0,0 +1,35 @@
+    model_part = Model[settings["model_part_name"].GetString()]
+    hdf5_file_factory = hdf5_io.HDF5SerialFileFactory(settings["file_settings"])
+    nodal_results_input = hdf5_io.NodalResultsInput(settings["nodal_results_settings"])
+    element_results_input = hdf5_io.ElementResultsInput(settings["element_results_settings"])
```
need to be modified according to new class "ElementDataValueInput"

sunethwarna commented on applications/HDF5Application/python_scripts/single_mesh_initialization_process.py:
```diff
@@ -0,0 +1,35 @@
+
+    initialization_process = hdf5_io.InitializationFromInputProcess(model_part,hdf5_file_factory,initialization_settings)
+    initialization_process.AddInput(nodal_results_input)
+    initialization_process.AddInput(element_results_input)
```
To be more consistent with the other processes, would it be ok to add nodal data value container initialization also?

msandre commented on applications/HDF5Application/python_scripts/single_mesh_temporal_input_process.py:
```diff
@@ -9,7 +9,9 @@ def Factory(settings, Model):
                 "model_part_name" : "MainModelPart",
-                "file_settings" : {},
+                "file_settings" : {
+                    "file_access_mode" : "read_only"
```
Can we omit "file_access_mode" here? The idea is to only define a default setting once in the code.

jcotela replied:
Are we sure that we want "read_only" to be the default setting? I think some truncate+write mode would make more sense as a default, since input processes are relatively few, but your call.

jcotela replied:
After looking into it, both single_mesh_temporal_input_process.py and single_mesh_initialization_process.py set the file access mode to read_only to override the "exclusive" default setting of the SerialFileFactory. Either process would fail if opening in exclusive mode (it throws an error if the file exists), so I'm afraid I need to set them both. Note that this is a modification of an existing process, which I now realize I should have pushed as a separate PR.

msandre replied:
I don't see `read_only` in any python script in the hdf5 app in master branch. Let's discuss it in person.

msandre commented on applications/HDF5Application/python_scripts/single_mesh_initialization_process.py:
```diff
@@ -0,0 +1,35 @@
+    nodal_results_input = hdf5_io.NodalResultsInput(settings["nodal_results_settings"])
+    element_results_input = hdf5_io.ElementResultsInput(settings["element_results_settings"])
+
+    initialization_settings = KratosMultiphysics.Parameters(r'''{}''')
```
Parameters should work with a default constructor now.

jcotela committed:
Updating method and argument names (consistent with output).

msandre commented on applications/HDF5Application/python_scripts/hdf5_io.py:
```diff
@@ -283,3 +283,31 @@ def _get_current_file_name(self):
         return self._time_step_file_name + time_tag + ".h5"
+
+class InitializationFromInputProcess(KratosMultiphysics.Process):
+    """A process to initialize a simulation using an HDF5 output file."""
```
I think this is like a static version of TemporalInputProcess, but it behaves slightly differently in how you parse the file_name. Can we adopt the same behaviour for parsing the file_name and call it something like StaticInputProcess or InitialInputProcess to be slightly more general?

msandre replied:
What I mean is to do

```
default_settings = KratosMultiphysics.Parameters("""{
        {
                 "file_name": "%s"
        }
    }""" % model_part.Name)
```

jcotela committed:
Review comments.

jcotela committed:
More review comments.

jcotela committed:
Removign unused process_info settings.

msandre commented on applications/HDF5Application/python_scripts/single_mesh_initialization_process.py:
```diff
@@ -0,0 +1,37 @@
+    initialization_settings.AddEmptyValue("file_name")
+    initialization_settings["file_name"].SetString(settings["file_name"].GetString())
+
+    initialization_process = hdf5_io.InitializationFromInputProcess(model_part,hdf5_file_factory,initialization_settings)
```
InitializationFromInputProcess has been renamed.

jcotela committed:
Merging current master.

RiccardoRossi commented:
guys, just a petition ... I think that this is potentially very very useful. if you may, take the time to add it to the wiki

jcotela committed:
Merging master.

jcotela committed:
Correcting file name.

jcotela committed:
Adding tests for temporal input and initialization processes.

jcotela committed:
Giving consistent default value for file_name in InitialInputProcess.

msandre commented on applications/HDF5Application/python_scripts/hdf5_io.py:
```diff
@@ -283,3 +283,37 @@ def _get_current_file_name(self):
+        KratosMultiphysics.Process.__init__(self)
+        default_settings = KratosMultiphysics.Parameters("""
+            {
+                "time_tag_precision": 4,
```
`time_tag_precision` is unused.

msandre commented on applications/HDF5Application/python_scripts/hdf5_io.py:
```diff
@@ -283,3 +283,37 @@ def _get_current_file_name(self):
+            """ % model_part.Name)
+        settings.ValidateAndAssignDefaults(default_settings)
+        file_name_string = settings["file_name"].GetString()
+        if file_name_string == "":
```
Now that we are giving model part name as default, there is no need to treat the empty string as a special case, i.e. we don't need this error check. In general, invalid filenames will generate exceptions from the file's constructor.

msandre commented on applications/HDF5Application/tests/test_hdf5_processes.py:
```diff
@@ -0,0 +1,181 @@
+
+        # Now initialize the read model part and compare the two model parts
+        input_process.ExecuteInitialize()
+
```
self.assertEqual(read_model_part.GetProperties(1)[VISCOSITY], write_model_part.GetProperties(1)[VISCOSITY])

msandre commented on applications/HDF5Application/tests/test_hdf5_processes.py:
```diff
@@ -0,0 +1,181 @@
+        input_settings = Parameters(r"""{
+            "Parameters" : {
+                "model_part_name" : "ReadModelPart",
+                "nodal_solution_step_data_settings" : {
```
Test would be more complete if we also included nodal and elemental data values.

msandre commented on applications/HDF5Application/tests/test_hdf5_processes.py:
```diff
@@ -0,0 +1,181 @@
+    def testSingleMeshTemporalOutputInitialization(self):
+        """
+        Output ModelPart using SingleMeshTemporalOutputProcess
+        Use output file to initialize (restart) a simulation with SingleMeshTemporalInputProcess
```
Bad comment.

msandre reviewed (changes requested):
I don't think the "single_mesh_" in single_mesh_initialization_process.py is necessary or meaningful in this context. The file name could be shortened to initialization_process.py or, if this is too general or creates a conflict for importing python scripts, then initialization_from_hdf5_process.py. "single_mesh" and "multiple_mesh" are used here to distinguish between hdf5 output with mesh information at each io (i.e., the  "multiple_mesh" case) versus one global mesh file for all io steps.

jcotela committed:
Merging master.

jcotela committed:
Renaming initialization process and updating comments.

jcotela committed:
Removing unused setting and leaving file name checks to File object.

jcotela committed:
Fixing new bug in process initialization and updating tests.

jcotela committed:
Checking values in elemental and non-historical nodal databases.

jcotela commented:
@msandre I believe I addressed the last round of issues. As discussed, I removed references to data in properties, since the processes involved in the tests do not read the ModelPart data file, only the result files.

msandre reviewed (approved):
Sorry for the wait. I'm still sorting though all the notifications that landed in my junk mail :mailbox_with_mail:  :hammer:
